ion-schema/src/system.rs

```diff
@@ -862,27 +929,65 @@ impl Resolver {
         // get all isl type names that are defined within the schema
         // this will be used to resolve type references which might not have yet resolved while loading a type definition
         let isl_type_names: HashSet<&str> = HashSet::from_iter(
-            isl.types()
+            isl_types
                 .iter()
-                .filter(|t| matches!(t, IslType::Named(_)))
+                .filter(|t| matches!(t, IslTypeKind::Named(_)))
                 .map(|t| t.name().as_ref().unwrap().as_str()),
         );
 
         // Resolve all ISL types and constraints
         for isl_type in isl.types() {
             let pending_types = &mut PendingTypes::default();
 
-            match isl_type {
-                IslType::Named(named_isl_type) => {
+            match &isl_type.kind {
+                IslTypeKind::Named(named_isl_type) => {
+                    match isl_version {
+                        IonSchemaLanguageVersion::V1_0 => {
+                            // verify if there are any constraints with ISL 2.0 for this isl_type
+                            let has_isl_2_0_constraints =
+                                isl_type.constraints().iter().any(|c| match c.constraint {
+                                    IslConstraintImpl::Regex(_)
+                                    | IslConstraintImpl::Element(_)
+                                    | IslConstraintImpl::Annotations(_)
+                                    | IslConstraintImpl::Scale(_) => {
+                                        c.version == IonSchemaLanguageVersion::V2_0
+                                    }
+                                    _ => false,
+                                });
```

Comment:
```suggestion
                                isl_type.constraints().iter().any(|c| c.version != IonSchemaLanguageVersion::V1_0);
```
It's much more readable if you just check the version of all of the constraints. (It's also fewer operations.)

Reply:
True. I will need to change the way constraints are constructed but I agree that this should be simpler.